File: frontend/src/games/blob/engine/rules.ts
// ブロブチェインのルール（落ち物パズル。同じ色を 4 つつなげると消え、連鎖するとおじゃまを送れる）。
// 時間（落下・アニメーション）は扱わない。controller.ts が時間に合わせてここの関数を呼ぶ。
//
// 盤面: 幅 6、高さ 13（y=0 が一番下。y=12 は画面外の 13 段目で、ここにある粒は消えない）。
// 組ぷよ（2 個 1 組）: 軸 (x, y) と子。rot 0 = 子が上、1 = 右、2 = 下、3 = 左。

export const W = 6
export const H = 13
export const VISIBLE_H = 12
export const SPAWN_X = 2
export const SPAWN_Y = 11 // この位置がふさがると負け（左から 3 列目・上から 2 段目）
export const EMPTY = 0
export const GARBAGE = 5
export const COLORS = 4
export const POP_COUNT = 4
export const TARGET_POINT = 70 // 何点で 1 個おじゃまを送るか
export const ALL_CLEAR_BONUS = 30 // 全消しした次の連鎖に上乗せするおじゃまの数
export const MAX_GARBAGE_DROP = 30 // 1 回に降るおじゃまの上限（5 段）

// 連鎖ボーナス（1 連鎖目から）・色数ボーナス・連結ボーナス（ぷよぷよ通と同じ計算）
const CHAIN_POWER = [0, 8, 16, 32, 64, 96, 128, 160, 192, 224, 256, 288, 320, 352, 384, 416, 448, 480, 512]
const COLOR_BONUS = [0, 0, 3, 6, 12, 24]
const GROUP_BONUS = (n: number) => (n <= 4 ? 0 : n >= 11 ? 10 : n - 3)

export interface Pair {
  x: number
  y: number
  rot: 0 | 1 | 2 | 3
  axis: number // 軸の色 1〜4
  child: number // 子の色
}

const CHILD_OFFSET: [number, number][] = [[0, 1], [1, 0], [0, -1], [-1, 0]]

export function childPos(p: Pair): [number, number] {
  const [dx, dy] = CHILD_OFFSET[p.rot]
  return [p.x + dx, p.y + dy]
}

export class Field {
  cells = new Uint8Array(W * H)

  get(x: number, y: number): number {
    if (x < 0 || x >= W || y < 0) return -1 // 壁と床
    if (y >= H) return EMPTY
    return this.cells[y * W + x]
  }

  set(x: number, y: number, v: number): void {
    if (x >= 0 && x < W && y >= 0 && y < H) this.cells[y * W + x] = v
  }

  free(x: number, y: number): boolean {
    return this.get(x, y) === EMPTY
  }

  isEmpty(): boolean {
    return this.cells.every((c) => c === EMPTY)
  }

  clone(): Field {
    const f = new Field()
    f.cells.set(this.cells)
    return f
  }

  /** 浮いている粒を下に落とす。戻り値は動いた粒の [x, 元の y, 新しい y]（アニメーション用） */
  applyGravity(): [number, number, number][] {
    const moved: [number, number, number][] = []
    for (let x = 0; x < W; x++) {
      let write = 0
      for (let y = 0; y < H; y++) {
        const v = this.get(x, y)
        if (v === EMPTY) continue
        if (y !== write) {
          this.set(x, write, v)
          this.set(x, y, EMPTY)
          moved.push([x, y, write])
        }
        write++
      }
    }
    return moved
  }

  /** 消える粒のまとまり（見えている段だけ・4 個以上）と、巻き込まれて消えるおじゃま */
  findPops(): { groups: number[][]; garbage: number[] } {
    const seen = new Uint8Array(W * H)
    const groups: number[][] = []
    for (let y = 0; y < VISIBLE_H; y++) {
      for (let x = 0; x < W; x++) {
        const c = this.get(x, y)
        const i = y * W + x
        if (c < 1 || c > COLORS || seen[i]) continue
        const group: number[] = []
        const stack = [i]
        seen[i] = 1
        while (stack.length) {
          const j = stack.pop()!
          group.push(j)
          const jx = j % W
          const jy = Math.floor(j / W)
          for (const [nx, ny] of [[jx + 1, jy], [jx - 1, jy], [jx, jy + 1], [jx, jy - 1]]) {
            if (nx < 0 || nx >= W || ny < 0 || ny >= VISIBLE_H) continue
            const k = ny * W + nx
            if (!seen[k] && this.cells[k] === c) {
              seen[k] = 1
              stack.push(k)
            }
          }
        }
        if (group.length >= POP_COUNT) groups.push(group)
      }
    }
    const garbage = new Set<number>()
    for (const g of groups) {
      for (const j of g) {
        const jx = j % W
        const jy = Math.floor(j / W)
        for (const [nx, ny] of [[jx + 1, jy], [jx - 1, jy], [jx, jy + 1], [jx, jy - 1]]) {
          if (nx >= 0 && nx < W && ny >= 0 && ny < VISIBLE_H && this.get(nx, ny) === GARBAGE) garbage.add(ny * W + nx)
        }
      }
    }
    return { groups, garbage: [...garbage] }
  }
}

/** 1 回の消去（連鎖の 1 段）の得点。chain は 1 から */
export function chainScore(groups: number[][], colorsOf: (g: number[]) => number, chain: number): number {
  const popped = groups.reduce((a, g) => a + g.length, 0)
  const colors = new Set(groups.map(colorsOf)).size
  const bonus = CHAIN_POWER[Math.min(chain - 1, CHAIN_POWER.length - 1)]
    + COLOR_BONUS[Math.min(colors, COLOR_BONUS.length - 1)]
    + groups.reduce((a, g) => a + GROUP_BONUS(g.length), 0)
  return 10 * popped * Math.min(999, Math.max(1, bonus))
}

// --- 乱数（seed が同じなら同じ順番。対戦では両者同じ組ぷよが来る） ---------------------------
export class Rng {
  private s: number
  constructor(seed: number) {
    this.s = seed >>> 0
  }
  next(): number {
    this.s = (this.s + 0x6d2b79f5) | 0
    let t = this.s
    t = Math.imul(t ^ (t >>> 15), t | 1)
    t = (t + Math.imul(t ^ (t >>> 7), t | 61)) ^ t
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296
  }
  int(n: number): number {
    return Math.floor(this.next() * n)
  }
}

File: backend/games/blob/rules.py
"""ブロブチェインのルール（TypeScript 版 frontend/src/games/blob/engine/rules.ts と同じ動きにする）。

時間（落下・アニメーション）は扱わない。盤面・組ぷよ・得点・乱数だけ。

盤面: 幅 6、高さ 13（y=0 が一番下。y=12 は画面外の 13 段目で、ここにある粒は消えない）。
組ぷよ（2 個 1 組）: 軸 (x, y) と子。rot 0 = 子が上、1 = 右、2 = 下、3 = 左。
"""

from __future__ import annotations

from dataclasses import dataclass, replace

W = 6
H = 13
VISIBLE_H = 12
SPAWN_X = 2
SPAWN_Y = 11  # この位置がふさがると負け（左から 3 列目・上から 2 段目）
EMPTY = 0
GARBAGE = 5
COLORS = 4
POP_COUNT = 4
TARGET_POINT = 70  # 何点で 1 個おじゃまを送るか
ALL_CLEAR_BONUS = 30  # 全消しした次の連鎖に上乗せするおじゃまの数
MAX_GARBAGE_DROP = 30  # 1 回に降るおじゃまの上限（5 段）

# 連鎖ボーナス（1 連鎖目から）・色数ボーナス・連結ボーナス（ぷよぷよ通と同じ計算）
CHAIN_POWER = [0, 8, 16, 32, 64, 96, 128, 160, 192, 224, 256, 288, 320, 352, 384, 416, 448, 480, 512]
COLOR_BONUS = [0, 0, 3, 6, 12, 24]

CHILD_OFFSET = [(0, 1), (1, 0), (0, -1), (-1, 0)]

# マスの隣（上下左右）を毎回作り直さずに済むように先に作っておく
_NEIGHBORS: list[tuple[int, ...]] = []
for _i in range(W * H):
    _x, _y = _i % W, _i // W
    _NEIGHBORS.append(tuple(
        ny * W + nx
        for nx, ny in ((_x + 1, _y), (_x - 1, _y), (_x, _y + 1), (_x, _y - 1))
        if 0 <= nx < W and 0 <= ny < VISIBLE_H
    ))


def group_bonus(n: int) -> int:
    return 0 if n <= 4 else 10 if n >= 11 else n - 3


@dataclass(frozen=True)
class Pair:
    """組ぷよ。axis が軸の色、child が子の色（どちらも 1〜4）。"""

    x: int
    y: int
    rot: int
    axis: int
    child: int

    def moved(self, **kw) -> "Pair":
        return replace(self, **kw)


def child_pos(p: Pair) -> tuple[int, int]:
    dx, dy = CHILD_OFFSET[p.rot]
    return p.x + dx, p.y + dy


class Field:
    """盤面。cells[y * W + x] に色（0 = 空き、1〜4 = 色、5 = おじゃま）。"""

    __slots__ = ("cells",)

    def __init__(self, cells: bytearray | None = None):
        self.cells = cells if cells is not None else bytearray(W * H)

    def get(self, x: int, y: int) -> int:
        if x < 0 or x >= W or y < 0:
            return -1  # 壁と床
        if y >= H:
            return EMPTY
        return self.cells[y * W + x]

    def set(self, x: int, y: int, v: int) -> None:
        if 0 <= x < W and 0 <= y < H:
            self.cells[y * W + x] = v

    def free(self, x: int, y: int) -> bool:
        return self.get(x, y) == EMPTY

    def is_empty(self) -> bool:
        return not any(self.cells)

    def clone(self) -> "Field":
        return Field(bytearray(self.cells))

    def height(self, x: int) -> int:
        """列 x に積まれている高さ（重力をかけた後は下から詰まっている）。"""
        h = 0
        while h < H and self.cells[h * W + x]:
            h += 1
        return h

    def heights(self) -> list[int]:
        return [self.height(x) for x in range(W)]

    def apply_gravity(self) -> list[tuple[int, int, int]]:
        """浮いている粒を下に落とす。戻り値は動いた粒の (x, 元の y, 新しい y)。"""
        moved: list[tuple[int, int, int]] = []
        cells = self.cells
        for x in range(W):
            write = 0
            for y in range(H):
                v = cells[y * W + x]
                if v == EMPTY:
                    continue
                if y != write:
                    cells[write * W + x] = v
                    cells[y * W + x] = EMPTY
                    moved.append((x, y, write))
                write += 1
        return moved

    def find_pops(self) -> tuple[list[list[int]], list[int]]:
        """消える粒のまとまり（見えている段だけ・4 個以上）と、巻き込まれて消えるおじゃま。"""
        cells = self.cells
        seen = bytearray(W * H)
        groups: list[list[int]] = []
        for i in range(W * VISIBLE_H):
            c = cells[i]
            if c < 1 or c > COLORS or seen[i]:
                continue
            group = [i]
            seen[i] = 1
            stack = [i]
            while stack:
                j = stack.pop()
                for k in _NEIGHBORS[j]:
                    if not seen[k] and cells[k] == c:
                        seen[k] = 1
                        stack.append(k)
                        group.append(k)
            if len(group) >= POP_COUNT:
                groups.append(group)
        garbage: list[int] = []
        hit = bytearray(W * H)
        for g in groups:
            for j in g:
                for k in _NEIGHBORS[j]:
                    if cells[k] == GARBAGE and not hit[k]:
                        hit[k] = 1
                        garbage.append(k)
        return groups, garbage


def can_place(field: Field, p: Pair) -> bool:
    """組ぷよ p をその場所に置けるか（壁・床・ほかの粒にぶつからないか）。"""
    cx, cy = child_pos(p)
    return field.free(p.x, p.y) and field.free(cx, cy) and p.y < H and cy < H


def pop_once(field: Field, chain: int) -> tuple[list[tuple[int, int]], int] | None:
    """消えるものがあれば 1 段分だけ消す（重力はまだかけない）。戻り値は (消えた (マス, 色), 得点)。

    BlobGame.pop_step と「連鎖が何段組めるか」の見積り（rl/blob/encoding.py）で共通に使う。
    """
    groups, garbage = field.find_pops()
    if not groups:
        return None
    cells = [(i, field.cells[i]) for i in [*(j for g in groups for j in g), *garbage]]
    score = chain_score(groups, lambda g: field.cells[g[0]], chain)
    for i, _ in cells:
        field.cells[i] = 0
    return cells, score


def simulate_chain(field: Field) -> tuple[int, int, int]:
    """連鎖が止まるまで進める（field は書き換わる）。戻り値は (連鎖数, 消えた数, 得点)。"""
    chain = popped = score = 0
    while True:
        step = pop_once(field, chain + 1)
        if step is None:
            return chain, popped, score
        cells, s = step
        chain += 1
        popped += len(cells)
        score += s
        field.apply_gravity()


def chain_score(groups: list[list[int]], color_of, chain: int) -> int:
    """1 回の消去（連鎖の 1 段）の得点。chain は 1 から。"""
    popped = sum(len(g) for g in groups)
    colors = len({color_of(g) for g in groups})
    bonus = (
        CHAIN_POWER[min(chain - 1, len(CHAIN_POWER) - 1)]
        + COLOR_BONUS[min(colors, len(COLOR_BONUS) - 1)]
        + sum(group_bonus(len(g)) for g in groups)
    )
    return 10 * popped * min(999, max(1, bonus))


class Rng:
    """mulberry32。TypeScript 版の Rng と同じ値を同じ順番で返す。"""

    __slots__ = ("s",)

    def __init__(self, seed: int):
        self.s = seed & 0xFFFFFFFF

    def next(self) -> float:
        # JavaScript の Math.imul（32bit の掛け算）は「掛けて下位 32bit を取る」と同じ
        self.s = (self.s + 0x6D2B79F5) & 0xFFFFFFFF
        t = self.s
        t = ((t ^ (t >> 15)) * (t | 1)) & 0xFFFFFFFF
        t = ((t + ((t ^ (t >> 7)) * (t | 61))) & 0xFFFFFFFF) ^ t
        return (t ^ (t >> 14)) / 4294967296

    def int(self, n: int) -> int:
        return int(self.next() * n)
